src/EditorFeatures/Core.Wpf/InlineHints/InlineHintsTagger.cs

```diff
@@ -126,86 +105,60 @@ private TextFormattingRunProperties Format
 
+                // If the snapshot has changed, we can't use any of the cached data, as it is associated with the
+                // original snapshot they were created against.
                 var snapshot = spans[0].Snapshot;
-                if (snapshot != _cacheSnapshot)
-                {
-                    // Calculate UI elements
-                    _cache.Clear();
-                    _cacheSnapshot = snapshot;
-
-                    // Calling into the InlineParameterNameHintsDataTaggerProvider which only responds with the current
-                    // active view and disregards and requests for tags not in that view
-                    var fullSpan = new SnapshotSpan(snapshot, 0, snapshot.Length);
-                    var tags = _tagAggregator.GetTags(new NormalizedSnapshotSpanCollection(fullSpan));
-                    foreach (var tag in tags)
-                    {
-                        // Gets the associated span from the snapshot span and creates the IntraTextAdornmentTag from the data
-                        // tags. Only dealing with the dataTagSpans if the count is 1 because we do not see a multi-buffer case
-                        // occurring
-                        var dataTagSpans = tag.Span.GetSpans(snapshot);
-                        if (dataTagSpans.Count == 1)
-                        {
-                            _cache.Add((tag, tagSpan: null));
-                        }
-                    }
-                }
 
                 var document = snapshot.GetOpenDocumentInCurrentContextWithChanges();
-                var classify = document != null && _taggerProvider.EditorOptionsService.GlobalOptions.GetOption(InlineHintsViewOptionsStorage.ColorHints, document.Project.Language);
+                if (document is null)
+                    return;
+
+                var colorHints = _taggerProvider.EditorOptionsService.GlobalOptions.GetOption(InlineHintsViewOptionsStorage.ColorHints, document.Project.Language);
+
+                using var _1 = SegmentedListPool.GetPooledList<TagSpan<InlineHintDataTag<CachedAdornmentTagSpan>>>(out var dataTagSpans);
+                _underlyingTagger.AddTags(spans, dataTagSpans);
+
+                // Presize so we can add the elements below without continually resizing.
+                adornmentTagSpans.Capacity += dataTagSpans.Count;
+
+                using var _2 = PooledHashSet<int>.GetInstance(out var seenPositions);
 
-                var selectedSpans = new List<TagSpan<IntraTextAdornmentTag>>();
-                for (var i = 0; i < _cache.Count; i++)
+                var format = this.Format;
+                foreach (var dataTagSpan in dataTagSpans)
                 {
-                    var tagSpans = _cache[i].mappingTagSpan.Span.GetSpans(snapshot);
-                    if (tagSpans.Count == 1)
-                    {
-                        var tagSpan = tagSpans[0];
-                        if (spans.IntersectsWith(tagSpan))
```

Comment:
we defer to the underlying tagger to do this.  as we passed along the span we were asked for to it, we don't have to do any additional processing.